src/main/java/ru/vk/itmo/osipovdaniil/DiskStorageUtils.java

```diff
@@ -0,0 +1,274 @@
+        return segment.get(ValueLayout.JAVA_LONG_UNALIGNED, 0);
+    }
+
+    static Iterator<Entry<MemorySegment>> iterator(final MemorySegment page,
+                                                   final MemorySegment from,
+                                                   final MemorySegment to) {
+        long recordIndexFrom = from == null ? 0 : normalize(indexOf(page, from));
+        long recordIndexTo = to == null ? recordsCount(page) : normalize(indexOf(page, to));
+        final long recordsCount = recordsCount(page);
+        return new Iterator<>() {
+            long index = recordIndexFrom;
+
+            @Override
+            public boolean hasNext() {
+                return index < recordIndexTo;
+            }
+
+            @Override
+            public Entry<MemorySegment> next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException();
+                }
+                MemorySegment key = slice(page, startOfKey(page, index), endOfKey(page, index));
+                long startOfValue = startOfValue(page, index);
+                MemorySegment value =
+                        startOfValue < 0
+                                ? null
+                                : slice(page, startOfValue, endOfValue(page, index, recordsCount));
+                index++;
+                return new BaseEntry<>(key, value);
+            }
+        };
+    }
+
+    private static MemorySegment slice(final MemorySegment page, final long start, final long end) {
+        return page.asSlice(start, end - start);
+    }
+
+    private static long startOfKey(final MemorySegment segment, final long recordIndex) {
+        return segment.get(ValueLayout.JAVA_LONG_UNALIGNED, recordIndex * 2 * Long.BYTES);
+    }
+
+    private static long endOfKey(final MemorySegment segment, final long recordIndex) {
+        return normalizedStartOfValue(segment, recordIndex);
+    }
+
+    private static long normalizedStartOfValue(final MemorySegment segment, final long recordIndex) {
+        return normalize(startOfValue(segment, recordIndex));
+    }
+
+    private static long startOfValue(final MemorySegment segment, final long recordIndex) {
+        return segment.get(ValueLayout.JAVA_LONG_UNALIGNED, recordIndex * 2 * Long.BYTES + Long.BYTES);
+    }
+
+    private static long endOfValue(final MemorySegment segment, final long recordIndex, final long recordsCount) {
+        if (recordIndex < recordsCount - 1) {
+            return startOfKey(segment, recordIndex + 1);
+        }
+        return segment.byteSize();
+    }
+
+    private static long tombstone(final long offset) {
+        return 1L << 63 | offset;
+    }
+
+    private static long normalize(final long value) {
+        return value & ~(1L << 63);
+    }
+
+    public static void save(final Path storagePath, final Iterable<Entry<MemorySegment>> iterable)
+            throws IOException {
+        final Path indexTmp = getIndexTmpPath(storagePath);
```

Comment:
[критикал]

Судя по коду в InMemoryDao
```java
if (iterable.iterator().hasNext()) {
            DiskStorageUtils.deleteAll(path);
            DiskStorageUtils.save(path, iterable);
        }
```
Файлы сначала удаляются, а потом вызывается save, где происходит чтение индекса и existedFiles, который ожидаемо пустой. Я добавил пару логов и убедился в этом

Данные в compaction файл пишутся из итераторов, которые в свою очередь замаплены удаленные файлы. Почему не бросает исключение  на тестах -- вопрос хороший, пока не могу на него ответить. 

Джавадок к FileChannel#map, который при вызове loadOrRecover инициализирует MemorySegment, с которым мы дальше работаем.


```
All or part of a mapped memory segment may become inaccessible at any time, for example if the backing mapped file is truncated. An attempt to access an inaccessible region of a mapped memory segment will not change the segment's content and will cause an unspecified exception to be thrown either at the time of the access or at some later time. It is therefore strongly recommended that appropriate precautions be taken to avoid the manipulation of a mapped file by this (or another) program, except to read or write the file's content.
```

Reply:
Согласен, пофиксил